scripts/cbmc-viewer/block.py

```diff
@@ -0,0 +1,274 @@
+        reachable) giving the number of reachable lines, the number of
+        hit lines, and their ratio coverage.
+    """
+
+    def __init__(self, xml="", srcdir="", blddir="",
+                 reachable=None, srcloc=None):
+        """Initialize block coverage statistics.
+
+        Args:
+
+           xml (str): the name of the xml file containing the block
+           coverage data produced by cbmc.
+
+           srcdir (str): the root of the source directory
+
+           blddir (str): the root of the source directory used to
+           build the goto binary (which may be different from srcdir
+           if the source directory was preprocessed before building)
+
+           reachable (obj): an instance of the Reachable class giving
+           statistics on the statically reachable functions
+        """
+
+        # pylint: disable=too-many-arguments
+
+        self.line = {}
+        self.function = {}
+        self.file = {}
+
+        # Record the list of reachable functions for later use
+        self.reachable = reachable.functions if reachable else []
+        # Record the mapping of filenames computed below for later use
+        # This should be using srcloc.clean_path
+        self.cleanup_filename = Block.cleanup_filename_function(blddir or
+                                                                srcdir)
+
+        if not xml:
+            return
+
+        root = SrcLoc.parse_xml_file(xml, abort=False)
+        if root is None:
+            return
+        srcloc.scan_source_locations(root)
+
+        line_status = Block.line_status(root, srcloc)
+        self.line = Block.line_coverage(line_status)
+        self.function = Block.function_coverage(line_status)
+        self.file = Block.file_coverage(line_status)
+
+    @staticmethod
+    def line_status(root, srcloc):
+        """Compute reachability of a line in a source file."""
+
+        # pylint: disable=too-many-locals
+
+        line_status = {}
+
+        for goal in root.iter("goal"):
+            goal_description = goal.get("description")
+            if not goal_description:
+                break
+            goal_status = goal.get("status")
+            goal_loc = goal.find("location")
+            loc_file = goal_loc.get("file")
+            loc_func = goal_loc.get("function")
+            loc_line = int(goal_loc.get("line"))
+
+            lnst = (linestatus.HIT
+                    if goal_status == "SATISFIED"
+                    else linestatus.MISSED)
+
+            locations = Block.parse_location_description(goal_description)
+            if not locations:
+                locations = [(loc_file, loc_func, loc_line)]
+
+            for loc in locations:
+                src, func, line = loc
+                src = srcloc.clean_path(src)
+
+                if not line_status.get(src):
+                    line_status[src] = {}
+                if not line_status[src].get(func):
```

Comment:
There are a lot of cases like this where it's "look this thing up in a map, if not initialise it to this, and then regardless of whether or not it was there before do this..." 

these should just be replaced with something like

```
line_status_file = line_status.setdefault(file, {})
line_status_file_function = line_status_file.setdefault(func, {})
```

etc, or otherwise be replace with utility classes with proper accessors that do that sort of bookkeeping.